packages/plans-grid-next/src/components/comparison-grid/hooks/useCompactViewOrder.ts

```diff
@@ -0,0 +1,87 @@
+import { useEffect } from 'react';
+import { GridPlan } from '../../../types';
+
+type Props = {
+	/**
+	 * Selected plan interval
+	 */
+	intervalType: UrlFriendlyTermType;
+	/**
+	 * The map of displayed plans for the current interval
+	 */
+	gridPlansIndex: { [ key: string ]: GridPlan };
+};
+
+/**
+ *
+ * When intervalType change trickles down to the ComparisonGrid as a prop (intervalType),
+ * we must detect this prop change and transform the current plan order into a set of plans
+ * that align with the currently selected plan interval without causing glitches or flashes.
+ *
+ * This hook serves as the solution. The hook accomplishes two main tasks:
+ * - Resolves the correct compact view plan order, providing it immediately in the current render cycle to prevent glitches.
+ * - Updates the WpcomPlansUI store with the most up-to-date plan order as required.
+ *
+ * ******
+ * @typedef {Object} Param1
+ * @property {UrlFriendlyTermType} intervalType - Selected plan interval
+ * @property {Object.<string, GridPlan>} gridPlansIndex - The map of displayed plans for the current interval
+ * @returns The list of gridPlans sorted by the accurate compact view order.
+ */
+
+function useCompactViewOrder( { intervalType, gridPlansIndex }: Props ): GridPlan[] {
+	const { setComparisonGridCompactViewPlanDisplayOrder } = useDispatch( WpcomPlansUI.store );
+	const compactViewPlanOrder = useSelect(
+		( select ) => select( WpcomPlansUI.store ).getComparisonGridCompactViewDisplayOrder(),
+		[]
+	);
+	const displayedPlanSlugs = Object.keys( gridPlansIndex ) as PlanSlug[];
+
+	let refreshedCompactViewPlanOrder: PlanSlug[] = compactViewPlanOrder;
+	let isCompactViewStale = false;
+	if ( compactViewPlanOrder.length > 0 ) {
+		// If at least one planSlug in the compactViewPlanOrder state is not present in the current displayedPlanSlugs,
+		// it indicates that the compactViewPlanOrder is now stale most probably due to a change in the interval.
+		isCompactViewStale = compactViewPlanOrder.some(
+			( plan ) => ! displayedPlanSlugs.includes( plan )
+		);
+
+		if ( isCompactViewStale ) {
+			refreshedCompactViewPlanOrder = compactViewPlanOrder.map( ( previousSlug ) => {
+				if ( isFreePlan( previousSlug ) ) {
+					return previousSlug;
+				}
+				return findFirstSimilarPlanKey( previousSlug, {
+					term: URL_FRIENDLY_TERMS_MAPPING[ intervalType ],
+				} ) as PlanSlug;
+			} );
+		}
+	} else {
+		// The WpcomPlansUI store is not initialized
+		isCompactViewStale = true;
+		refreshedCompactViewPlanOrder = displayedPlanSlugs;
+	}
+	useEffect( () => {
+		if ( isCompactViewStale ) {
+			setComparisonGridCompactViewPlanDisplayOrder( { planOrder: refreshedCompactViewPlanOrder } );
+		}
+	}, [
+		isCompactViewStale,
+		refreshedCompactViewPlanOrder,
+		setComparisonGridCompactViewPlanDisplayOrder,
+	] );
+	return refreshedCompactViewPlanOrder
+		.map( ( planSlug ) => gridPlansIndex[ planSlug ] )
+		.filter( ( i ) => i );
+}
+
+export default useCompactViewOrder;
```

Comment:
Honestly, this feels complex to me. There is a lot going on here, which I'd prefer we revise the approach at a higher level first, before diving into the details.